modules/@apostrophecms/command-menu/index.js

```diff
@@ -0,0 +1,310 @@
+
+        return {
+          ...initialState,
+          modals: concatenate || {}
+        };
+      },
+      validateCommand({ name, command }) {
+        try {
+          assert.equal(command.type, 'item', `Invalid command type, must be "item", for ${name}`);
+          assert.equal(typeof command.label, 'string', `Invalid command label, must be a string, for ${name} "${typeof command.label}" provided`);
+          assert.equal(typeof command.action, 'object', `Invalid command action, must be an object for ${name}`) &&
+            assert.equal(typeof command.action.type, 'string', `Invalid command action type for ${name}`) &&
+            assert.equal(typeof command.action.payload, 'object', `Invalid command action payload for ${name}`);
+          command.permission && (
+            assert.equal(typeof command.permission, 'object', `Invalid command permission for ${name}`) &&
+            assert.equal(typeof command.permission.action, 'string', `Invalid command permission action for ${name}`) &&
+            assert.equal(typeof command.permission.type, 'string', `Invalid command permission type for ${name}`)
+          );
+          command.modal &&
+            assert.equal(typeof command.modal, 'string', `Invalid command modal for ${name}`);
+          assert.equal(typeof command.shortcut, 'string', `Invalid command shortcut, must be a string, for ${name}`);
+
+          return [ true, null ];
+        } catch (error) {
+          return [ false, error ];
+        }
+      },
+      validateGroup({ name, group }) {
+        try {
+          group.label && typeof group.label === 'object'
+            ? assert.equal(typeof group.label.key, 'string', `Invalid group label key for ${name}`)
+            : assert.equal(typeof group.label, 'string', `Invalid group label, must be a string, for ${name} "${typeof group.label}" provided`);
+          assert.equal(Array.isArray(group.commands), true, `Invalid group commands, must be an array for ${name}`);
+          assert.ok(group.commands.every(field => typeof field === 'string'), `Invalid group commands, must contains strings, for ${name}`);
+
+          return [ true, null ];
+        } catch (error) {
+          return [ false, error ];
+        }
+      },
+      validateModal({ name, modal }) {
+        return Object.entries(modal)
+          .map(([ groupName, group ]) => self.validateGroup({
+            name: `${name}:${groupName}`,
+            group
+          }));
+      },
+      compileErrors(result) {
+        const errors = result
+          .filter(([ success ]) => !success)
+          .map(([ , error ]) => error);
+        if (errors.length) {
+          throw new Error('Invalid', { cause: errors });
+        }
+      },
+      buildCommands(initialState) {
+        const concatenate = self.apos.util.omit(
+          initialState.composed.commands,
+          initialState.composed.removes
```

Comment:
This needs to respect inheritance order. If a base class adds a command, a subclass removes it, and a sub-subclass adds it back  (often project level for instance), then that needs to work. Make sure you include that unit test.

That means this has to happen iteratively through the layers of the initialization of the object — each layer must add, remove and group things, and then the next layer has a go at it.

Look at how it is done in `moog.js` where the regular cascades are implemented. If you can't just make this a regular cascade and benefit from that logic like fields, columns, etc. do (could you? Worth a look) then you need to make sure you have that same capability.

Reply:
I've achieved that by building sequential removes in https://github.com/apostrophecms/apostrophe/pull/3935/files#diff-91dd88c4b9d7b9798c4f0b88ef58039a9adbd9383ec8b24a035694d85a344a91R79

and removing group commands duplicates in https://github.com/apostrophecms/apostrophe/pull/3935/files#diff-91dd88c4b9d7b9798c4f0b88ef58039a9adbd9383ec8b24a035694d85a344a91R113

As discussed yesterday, `cascades` won't achieve my goal because I'm building against a single pool of data shared across all the apostrophe instance instead of one pool of data per module like we do with moog.